### PR TITLE
Don't convert value if it is already a valid type.

### PR DESCRIPTION
**Pull request by KotlinIsland**
#3867
#3738

If a value is already a valid type from a type union then there is no need to convert it.

### CODE REVIEW

**Comment by KotlinIsland** [comment]
Or would this be better in `ArgumentConverter._convert`?

**Comment by pekkaklarck** [comment]
The current functionality is there by design. The motivation is that we want conversion to be done if you have type like `Union[int, str]` and use argument like `42`. I don't want that to be changed, but I don't like too much that currently if you have `Union[int, float]` and use it with `${42.9}` that gets converted to an integer `42`. Possibly we should always convert strings but with others with check is the type valid already. Anyway, this change needs a separate issue and cannot be done before RF 4.1.

**Comment by KotlinIsland** [comment]
The current functionality breaks optional strings `Optional[str]` completely.

If the purpose of this is to allow int/float/enum literals to override `str` then can we skip forced conversion when value is not a string

```py
    def _no_conversion_needed(self, value):
        if isinstance(value, str):
            return False
        # None is not a valid type, convert to NoneType
        valid_instance_tuple = tuple(type(it) if it is None else it for it in self.args)
        return isinstance(value, valid_instance_tuple)
```

Or just always put `str` last in the conversion? I don't like this option as it's adding more complexity for the user.
```
    def _convert(self, value, explicit_type=True):
        if type(str) in self.args:
            args = tuple(arg for arg in self.args if arg != type(str)) + type(str)
        else:
            args = self.args
        for typ in args:
            converter = TypeConverter.converter_for(typ)
            if not converter:
                return value
            try:
                return converter.convert('', value, explicit_type)
            except ValueError:
                pass
        raise ValueError
```

**Comment by pekkaklarck** [comment]
How does the current functionality break `Optional[str]`? This anyway needs an issue to discuss the design.

**Comment by KotlinIsland** [comment]
`Optional[str]` becomes `Union[str, None]` making it impossible to pass `None` from robot.

**Comment by KotlinIsland** [comment]
issue: #3897

**Comment by pekkaklarck** [comment]
I agree `${None}` being converted to a string is not good if you use `Optional[str]`, but I wouldn't say that totally breaks it. There's also a simple workaround that you omit `Optional` and just use `arg=None`. Anyway, let's discuss this more in #3897.

**Comment by pekkaklarck** [comment]
After thinking this more, I believe the approach this PR originally used is the best one. See [my comment](https://github.com/robotframework/robotframework/issues/3897#issuecomment-808336570) to issue #3897 for more details. I also submitted a separate issue about the problem with `arg: Optional[str] = None` (#3908).

File: src/robot/running/arguments/typeconverters.py [hunk 1]
```diff
@@ -451,7 +451,11 @@ def _handles_value(self, value):
         return True
 
     def _no_conversion_needed(self, value):
-        return False
+        if isinstance(value, str):
+            return False
+        # None is not a valid type, convert to NoneType
+        valid_instance_tuple = tuple(type(it) if it is None else it for it in self.args)
```
Comment (pekkaklarck) [inline src/robot/running/arguments/typeconverters.py:457]: This could be written like
```python
NoneType = type(None)
types = tuple(typ if typ is not None else NoneType for typ in self.args)
```
to avoid the need for the comment.

**Comment by pekkaklarck** [comment]
Looks great. Docs related to Union conversion need to still be updated but I can do that myself. I'll also make sure we have tests covering the `arg: Optional[str] = None` case (#3908).

**Comment by pekkaklarck** [comment]
One of the tests was actually failing. I should've checked CI status more before merging, but I was anyway planning to enhance tests a bit so I'll fix the test at the same time.
